torch/csrc/jit/export.cpp

```diff
@@ -638,6 +642,186 @@ void ScriptModuleSerializer::writeTensorTable(torch::ModelDef* model_def) {
+      AT_ERROR("Unsupported IValue type for pickling (TensorList)");
+    } else if (ivalue.isGenericList()) {
+      pushList(ivalue);
+    } else if (ivalue.isGenericDict()) {
+      pushDict(ivalue);
+    } else if (ivalue.isObject()) {
+      AT_ERROR("Unsupported IValue type for pickling (Object)");
+    } else if (ivalue.isNone()) {
+      push(OpCode::NONE);
+    } else if (ivalue.isDevice()) {
+      AT_ERROR("Unsupported IValue type for pickling (Device)");
+    } else {
+      AT_ERROR("Unknown IValue type for pickling");
+    }
+
+    return 0;
+  }
+
+private:
+  void pushString(const IValue& ivalue) {
+    auto string = ivalue.toStringRef();
+
+    push(OpCode::BINUNICODE);
+    push(uint32_t(string.size()));
+    stack_.insert(stack_.end(), string.begin(), string.end());
+
+    push(OpCode::BINPUT, memo_id++);
+  }
+
+  void pushDouble(const IValue& ivalue) {
+    double value = ivalue.toDouble();
+    AT_ASSERT(sizeof(double) == 8);
+    char bytes[8];
+    double* bytes_as_double = reinterpret_cast<double*>(bytes);
+    *bytes_as_double = value;
+
+    push(OpCode::BINFLOAT);
+    for (size_t i = 0; i < sizeof(bytes); ++i) {
+      push(bytes[sizeof(bytes) - i - 1]);
+    }
+  }
+
+  void pushDict(const IValue& ivalue) {
+    auto dict = ivalue.toGenericDictRef();
+
+    push(OpCode::EMPTY_DICT);
+    push(OpCode::BINPUT, memo_id++);
+
+    push(OpCode::MARK);
+
+    for (auto pair : dict) {
+      addIValue(pair.first);
+      addIValue(pair.second);
+    }
+
+    push(OpCode::SETITEMS);
+  }
+
+  void pushList(const IValue& ivalue) {
+    auto list = ivalue.toGenericListRef();
+    push(OpCode::EMPTY_LIST);
+    push(OpCode::BINPUT, memo_id++);
+
+    push(OpCode::MARK);
+
+    for (auto item : list) {
+      addIValue(item);
+    }
+
+    push(OpCode::APPENDS);
+  }
+
+  void pushTuple(const IValue& ivalue) {
+    // TODO: Small tuple unrolling (e.g. TUPLE3)
+    push(OpCode::MARK);
+
+    for (const auto& item : ivalue.toTuple()->elements()) {
+      addIValue(item);
+    }
+
+    push(OpCode::TUPLE);
+    push(OpCode::BINPUT, memo_id++);
+  }
+
+  size_t pushBytes(size_t num_bytes) {
+    AT_ASSERT(num_bytes > 0);
+    stack_.push_back(0);
+    auto index = stack_.size() - 1;
+
+    for (size_t i = 1; i < num_bytes; ++i) {
+      stack_.push_back(0);
+    }
+    return index;
+  }
+
+  template <typename T>
+  size_t push(T value) {
+    auto index = pushBytes(sizeof(T));
+    T* ptr = reinterpret_cast<T*>(&stack_[index]);
+    *ptr = value;
+    return index;
+  }
+
+  template <typename T>
+  size_t push(OpCode opcode, T value) {
+    push(opcode);
+    auto index = pushBytes(sizeof(T));
+    T* ptr = reinterpret_cast<T*>(&stack_[index]);
+    *ptr = value;
+    return index;
+  }
+
+  // Stack of opcodes/data
+  std::vector<char> stack_;
+
+  // Memoization of IValues that have been written (index in table is used for
+  // BINPUT opcodes) to enable shared references
+  std::vector<IValue> memo_;
```

Comment:
This is unused. I think a std::unordered_map<void*, size_t> is actually what is needed here (pointer to memo_id if already saved).